canister/src/unstable_blocks.rs

```diff
@@ -71,6 +75,48 @@ impl UnstableBlocks {
     pub fn blocks_with_depths_by_heights(&self) -> Vec<Vec<(&Block, u32)>> {
         self.tree.blocks_with_depths_by_heights()
     }
+
+    /// Returns depth in BlockTree of Block with given BlockHash.
+    pub fn block_depth(&mut self, block_hash: &BlockHash) -> Result<u32, BlockDoesNotExtendTree> {
+        let (_, depth) = blocktree::find_mut(&mut self.tree, block_hash)
+            .ok_or_else(|| BlockDoesNotExtendTree(block_hash.clone()))?;
+        Ok(depth)
+    }
+
+    // Inserts the block hash of the block that should be received.
+    pub fn insert_next_block(
+        &mut self,
+        prev_block_hash: &BlockHash,
+        block_hash: &BlockHash,
+        stable_height: Height,
+    ) {
+        let height = match self.next_blocks.get_height(prev_block_hash) {
+            Some(prev_height) => *prev_height,
+            None => {
+                if let Ok(depth) = self.block_depth(prev_block_hash) {
+                    stable_height + depth
+                } else {
+                    return;
+                }
+            }
+        } + 1;
+        self.next_blocks.insert(block_hash, height);
+    }
+
+    // Removes the received block from 'next_blocks'.
+    pub fn remove_received_block_from_next_blocks(&mut self, received_block: &BlockHash) {
+        self.next_blocks.remove_block(received_block);
+    }
+
+    // Removes blocks from 'next_blocks' with height up to current 'stable_height'.
+    pub fn remove_next_blocks_based_on_stable_height(&mut self, stable_height: Height) {
+        self.next_blocks.remove_until_height(stable_height);
+    }
+
+    // Public only for testing purpose.
+    pub fn next_blocks_max_height(&self) -> Height {
+        self.next_blocks.get_max_height().unwrap_or(0)
```

Comment:
I find it odd that we're returning a height of zero here when there are no next blocks. Wouldn't be more clear to the callers if this returned an `Option<Height>`.